keras/applications/densenet.py

```diff
@@ -0,0 +1,333 @@
+            layer at the top of the network.
+        weights: one of `None` (random initialization),
+              'imagenet' (pre-training on ImageNet),
+              or the path to the weights file to be loaded.
+        input_tensor: optional Keras tensor (i.e. output of `layers.Input()`)
+            to use as image input for the model.
+        input_shape: optional shape tuple, only to be specified
+            if `include_top` is False (otherwise the input shape
+            has to be `(224, 224, 3)` (with `channels_last` data format)
+            or `(3, 224, 224)` (with `channels_first` data format).
+            It should have exactly 3 inputs channels.
+        pooling: optional pooling mode for feature extraction
+            when `include_top` is `False`.
+            - `None` means that the output of the model will be
+                the 4D tensor output of the
+                last convolutional layer.
+            - `avg` means that global average pooling
+                will be applied to the output of the
+                last convolutional layer, and thus
+                the output of the model will be a 2D tensor.
+            - `max` means that global max pooling will
+                be applied.
+        classes: optional number of classes to classify images
+            into, only to be specified if `include_top` is True, and
+            if no `weights` argument is specified.
+
+    # Returns
+        A Keras model instance.
+
+    # Raises
+        ValueError: in case of invalid argument for `weights`,
+            or invalid input shape.
+    """
+    if not (weights in {'imagenet', None} or os.path.exists(weights)):
+        raise ValueError('The `weights` argument should be either '
+                         '`None` (random initialization), `imagenet` '
+                         '(pre-training on ImageNet), '
+                         'or the path to the weights file to be loaded.')
+
+    if weights == 'imagenet' and include_top and classes != 1000:
+        raise ValueError('If using `weights` as imagenet with `include_top`'
+                         ' as true, `classes` should be 1000')
+
+    # Determine proper input shape
+    input_shape = _obtain_input_shape(input_shape,
+                                      default_size=224,
+                                      min_size=224,
+                                      data_format=K.image_data_format(),
+                                      require_flatten=include_top,
+                                      weights=weights)
+
+    if input_tensor is None:
+        img_input = Input(shape=input_shape)
+    else:
+        if not K.is_keras_tensor(input_tensor):
+            img_input = Input(tensor=input_tensor, shape=input_shape)
+        else:
+            img_input = input_tensor
+
+    bn_axis = 3 if K.image_data_format() == 'channels_last' else 1
+
+    x = ZeroPadding2D(padding=((3, 3), (3, 3)))(img_input)
```

Comment:
You don't need this, you can just use `padding='same'` in the layer below.

Reply:
@fchollet, Theano and Tensorflow produced different results without `ZeroPadding2D`.
I tested the following 4 cases with the below codes:
- explicit `ZeroPadding2D` followed by `Conv2D(padding='valid')` under Theano
- `Conv2D(padding='same')` under Theano
- explicit `ZeroPadding2D` followed by `Conv2D(padding='valid')` under TensorFlow
- `Conv2D(padding='same')` under TensorFlow

As a result, only the last item shows a different output. In other words, Theano and TensorFlow can have the same outputs (currently) only if `ZeroPadding2D` is added explicitly.
```python
%env KERAS_BACKEND=tensorflow
import numpy as np
from keras.models import Model
from keras.layers import Input, Conv2D, ZeroPadding2D

inputs = Input(shape=(224, 224, 1))

x1 = ZeroPadding2D(padding=(3, 3))(inputs)
x1 = Conv2D(1, 7, strides=2, use_bias=False)(x1)
model1 = Model(inputs, x1)

x2 = Conv2D(1, 7, strides=2, use_bias=False, padding='same')(inputs)
model2 = Model(inputs, x2)

w = [np.arange(49).reshape((7, 7, 1, 1))]
if K.backend() == 'tensorflow':
    w = [np.flip(np.flip(w[0], 0), 1)]
model1.set_weights(w)
model2.set_weights(w)

x = np.ones((1, 224, 224, 1))
print(model1.predict(x)[0,:3,:3,0])
print(model2.predict(x)[0,:3,:3,0])
```